src/test/rgw/test_d4n_filter.cc

```diff
@@ -0,0 +1,441 @@
+      cfg.store_name = "dbstore";
+      cfg.filter_name = "d4n";
+
+      auto filterDriver = DriverManager::get_raw_storage(env->dpp, g_ceph_context,
+							  cfg, io, site_config);
+
+      rgw::sal::Driver* next = filterDriver;
+      driver = newD4NFilter(next, io);
+    } 
+
+    virtual void TearDown() {
+      delete conn;
+    }
+
+    net::io_context io;
+    connection* conn; 
+
+    rgw::sal::Driver* driver;
+    std::unique_ptr<rgw::sal::Object> obj;
+    std::unique_ptr<rgw::sal::User> testUser = nullptr;
+    std::unique_ptr<rgw::sal::Bucket> testBucket = nullptr;
+    std::unique_ptr<rgw::sal::Writer> testWriter = nullptr;
+};
+
+void init_driver(rgw::sal::Driver* driver, boost::asio::yield_context yield) {
+  auto d4nFilter = dynamic_cast<rgw::sal::D4NFilterDriver*>(driver);
+  dynamic_cast<rgw::d4n::LFUDAPolicy*>(d4nFilter->get_policy_driver()->get_cache_policy())->save_y(optional_yield{yield});
+  driver->initialize(env->cct.get(), env->dpp);
+
+  ASSERT_NE(driver, nullptr);
+}
+
+void create_user(rgw::sal::Driver* driver, std::unique_ptr<rgw::sal::User>* testUser, boost::asio::yield_context yield) {
+  rgw_user u("test_tenant", "test_user", "ns");
+
+  *testUser = driver->get_user(u);
+  (*testUser)->get_info().user_id = u;
+
+  ASSERT_EQ((*testUser)->store_user(env->dpp, optional_yield{yield}, false), 0);
+}
+
+void create_bucket(std::string name, rgw::sal::Driver* driver, std::unique_ptr<rgw::sal::Bucket>* testBucket, boost::asio::yield_context yield) {
+  rgw::sal::Bucket::CreateParams createParams;
+  rgw_bucket b;
+  init_bucket(&b, "test_tenant", "test_name", "test_data_pool", "test_index_pool", "test_marker", "test_id");
+
+  EXPECT_EQ(driver->load_bucket(env->dpp, b, testBucket, optional_yield{yield}), 0);
+  ASSERT_EQ((*testBucket)->create(env->dpp, createParams, optional_yield{yield}), 0);
+  (*testBucket)->get_info().bucket.bucket_id = "test_bucket_" + name;
+}
+
+void put_object(std::string name, std::unique_ptr<rgw::sal::Object>* obj, rgw::sal::Driver* driver, std::unique_ptr<rgw::sal::Bucket>* testBucket, 
+                 std::unique_ptr<rgw::sal::Writer>* testWriter, boost::asio::yield_context yield) {
+  std::string object_name = "test_object_" + name;
+  *obj = (*testBucket)->get_object(rgw_obj_key(object_name));
+  ASSERT_NE((*obj).get(), nullptr);
+
+  rgw_owner owner;
+  ACLOwner acl_owner;
+  rgw_placement_rule ptail_placement_rule;
+  uint64_t olh_epoch = 123;
+  std::string unique_tag;
+
+  rgw_user uid{"test_tenant", "test_filter"};
+  owner = uid;
+  acl_owner.id = owner; 
+  *testWriter = driver->get_atomic_writer(env->dpp, 
+					  optional_yield{yield},
+					  (*obj).get(),
+					  acl_owner,
+					  &ptail_placement_rule,
+					  olh_epoch,
+					  "test_filter");
+
+  size_t accounted_size = 4;
+  std::string etag("test_etag");
+  ceph::real_time mtime; 
+  ceph::real_time set_mtime;
+
+  buffer::list bl;
+  bl.append("test_version");
+  rgw::sal::Attrs attrs{{RGW_ATTR_ID_TAG, bl}};
```

Comment:
why is the ID_TAG being passed in as an attribute? does dbstore not return ID_TAG as part of the attrs when read from the backend.

Reply:
I was having trouble with retrieving the ID_TAG during initial debugging, so I explicitly wrote it to the attrs to follow the value more closely. Since I changed `calculate_version()`, I will remove this and see if dbstore provides one instead.

Reply:
ok, see comment above